src/interface/web/app/components/loginPrompt/loginPrompt.tsx

```diff
@@ -1,47 +1,536 @@
+            </Drawer>
+        );
+    }
+
+    return (
+        <Dialog open={true} onOpenChange={props.onOpenChange}>
+            <DialogContent
+                className={`flex flex-col gap-4 ${!useEmailSignIn ? "p-0 pb-4 m-0 max-w-xl" : "w-fit"}`}
+            >
+                <div>
+                    {useEmailSignIn && (
+                        <EmailSignInContext
+                            email={email}
+                            setEmail={setEmail}
+                            checkEmail={checkEmail}
+                            setCheckEmail={setCheckEmail}
+                            setUseEmailSignIn={setUseEmailSignIn}
+                            recheckEmail={recheckEmail}
+                            setRecheckEmail={setRecheckEmail}
+                            handleMagicLinkSignIn={handleMagicLinkSignIn}
+                        />
+                    )}
+                    {!useEmailSignIn && (
+                        <MainSignInContext
+                            handleGoogleScriptLoad={handleGoogleScriptLoad}
+                            handleGoogleSignIn={handleGoogleSignIn}
+                            isLoading={isLoading}
+                            data={data}
+                            setUseEmailSignIn={setUseEmailSignIn}
+                            isMobileWidth={props.isMobileWidth ?? false}
+                        />
+                    )}
+                </div>
+            </DialogContent>
+        </Dialog>
+    );
+}
+
+function EmailSignInContext({
+    email,
+    setEmail,
+    checkEmail,
+    setCheckEmail,
+    setUseEmailSignIn,
+    recheckEmail,
+    handleMagicLinkSignIn,
+}: {
+    email: string;
+    setEmail: (email: string) => void;
+    checkEmail: boolean;
+    setCheckEmail: (checkEmail: boolean) => void;
+    setUseEmailSignIn: (useEmailSignIn: boolean) => void;
+    recheckEmail: boolean;
+    setRecheckEmail: (recheckEmail: boolean) => void;
+    handleMagicLinkSignIn: () => void;
+}) {
+    const [otp, setOTP] = useState("");
+    const [otpError, setOTPError] = useState("");
+    const [numFailures, setNumFailures] = useState(0);
+
+    function checkOTPAndRedirect() {
+        const verifyUrl = `/auth/magic?code=${otp}&email=${email}`;
+
+        if (numFailures >= ALLOWED_OTP_ATTEMPTS) {
+            setOTPError("Too many failed attempts. Please try again tomorrow.");
+            return;
+        }
+
+        fetch(verifyUrl, {
+            method: "GET",
+            headers: {
+                "Content-Type": "application/json",
+            },
+        })
+            .then((res) => {
+                if (res.ok) {
+                    // Check if the response is a redirect
+                    if (res.redirected) {
+                        window.location.href = res.url;
+                    }
+                } else if (res.status === 401) {
+                    setOTPError("Invalid OTP.");
+                    setNumFailures(numFailures + 1);
```

Comment:
This error does not differentiate between expired OTP versus incorrectly entered OTP. The two scenarios required different actions by the user (resent OTP email vs verify OTP entered by them is correct). This will require returning a different auth error from the API imo